### PR TITLE
[SG2] Refactor Command Handler registration and Graph load handling + Preview Scene Cleanup

### PR DESCRIPTION
**Pull request by Nightmask3**
---
### Purpose of this PR

Cleans up and breaks apart the monolithic command handler registration we were doing prior to this into more reasonable chunks across the tool.

Before, everything was handled in one single place, in the `Update` of the `ShaderGraphEditorWindow`

image

This is bad because:

1) We are not differentiating between the how command handlers are bound to certain views and their dispatchers, which is very important to understand
2) By lumping everything into a single function call, it is difficult to reason about what resources are actually needed by each subset of command handlers (eg. Overrides of GTF commands vs commands unique to shader graph)
3) Makes maintenance a pain by virtue of being difficult to reason about, and will only grow more annoying as more command handlers are added to that giant monolithic list

This PR improves this by:

Breaking up command handlers into:

- Overrides of GTF commands for the graph view (node and edge interactions) -> Now handled in `ShaderGraphView::GraphViewCommandOverridesRegistrar`
- Overrides of GTF commands for the blackboard -> Now handled in `ShaderGraphEditorWindow::CreateBlackboardView`
- Shader Graph commands that are unique to our tool -> Now handled in `ShaderGraphEditorWindow::CreateGraphTool` and `ShaderGraphEditorWindow::HandleGraphLoad` (the last is only for preview commands, which require the `GraphModel` to be loaded)

And it adds `ShaderGraphLoadedObserver` which we can rely on now for a concrete point at which the `ShaderGraphModel` is loaded and ready, at which point it calls `ShaderGraphEditorWindow::HandleGraphLoad` 

Also this handles cleanup of preview scene resources which wasn't done before.

---
### Testing status
Describe what manual/automated tests were performed for this PR

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.sg2/Editor/GraphDeltaRegistry/PreviewManager/HeadlessPreviewManager.cs [hunk 1]
```diff
@@ -121,19 +121,40 @@ static Texture2D GenerateFourSquare(Color c1, Color c2)
...
+            }
+
+            if (m_SceneResources != null)
+            {
+                m_SceneResources.Dispose();
+                m_SceneResources = null;
+            }
+
+            m_PreviewMaterialPropertyBlock.Clear();
+        }
```
Comment (esmelusina) [inline com.unity.sg2/Editor/GraphDeltaRegistry/PreviewManager/HeadlessPreviewManager.cs:156]: Making a note to ask if per-preview cleanup also happens as resources are no longer.

**Review by esmelusina — APPROVED** [review]
Seems fine-- get a second opinion.

It can be useful to add comments into the PR for what's new vs what is a move.

File: com.unity.sg2/Editor/GraphDeltaRegistry/PreviewManager/HeadlessPreviewManager.cs (hunk 1 above)
Comment (Nightmask3) [reply com.unity.sg2/Editor/GraphDeltaRegistry/PreviewManager/HeadlessPreviewManager.cs:156]: good point, there's a few more resources I need to cleanup here, most importantly the map of all the per-preview data 👍

File: com.unity.sg2/Editor/GraphDeltaRegistry/PreviewManager/HeadlessPreviewManager.cs [hunk 2]
```diff
@@ -121,19 +121,40 @@ static Texture2D GenerateFourSquare(Color c1, Color c2)
...
+                m_ErrorTexture = null;
+            }
+
+            if (m_SceneResources != null)
+            {
+                m_SceneResources.Dispose();
+                m_SceneResources = null;
+            }
+
+            m_PreviewMaterialPropertyBlock.Clear();
```
[the remark this answers is not in the document]
Comment (Nightmask3) [reply com.unity.sg2/Editor/GraphDeltaRegistry/PreviewManager/HeadlessPreviewManager.cs:155]: hmm when you say runtime cleanup, clarify for me as I dont understand

File: com.unity.sg2/Editor/GraphUI/EditorCommon/Preview/PreviewManager.cs [hunk 3]
```diff
@@ -181,6 +153,48 @@ public void Update()
...
+                    updatedNodes.Add(nodeName);
+                    break;
+                case HeadlessPreviewManager.PreviewOutputState.ShaderError:
+                {
+                    updatedNodes.Add(nodeName);
+
+                    // TODO: Handle displaying main preview errors
+                    foreach (var errorMessage in shaderMessages)
+                    {
+                        Debug.Log(errorMessage);
```
[the remark this answers is not in the document]
Comment (Nightmask3) [reply com.unity.sg2/Editor/GraphUI/EditorCommon/Preview/PreviewManager.cs:176]: Oh right yeah that's a good point I sort of just threw this is here a while back and didn't think more of it, I assumed that the implicit ToString() would work out, made that change 👍

File: com.unity.sg2/Editor/GraphUI/GraphElements/Views/ShaderGraphView.cs [hunk 4]
```diff
@@ -2,37 +2,103 @@
...
             if(window != null)
                 ViewSelection = new ShaderGraphViewSelection(this, GraphViewModel.GraphModelState, GraphViewModel.SelectionState);
+
+            GraphViewCommandOverridesRegistrar(this, graphTool.State, m_PreviewManager);
+        }
+
+        /// <summary>
+        /// Place to register any commands that are overrides of base GTF commands
+        /// </summary>
+        static void GraphViewCommandOverridesRegistrar(ShaderGraphView graphView, IState stateStore, PreviewManager previewManager)
```
[the remark this answers is not in the document]
Comment (Nightmask3) [reply com.unity.sg2/Editor/GraphUI/GraphElements/Views/ShaderGraphView.cs:48]: Oh yeah thats a good point too, I borrowed the `Registrar` terminology from GTF cause they seem to have this convention of static classes that handle all of their command binding, which we could also do, but our need to register and use command handlers is less extensive than what they need

I'll change the class name of `ShaderGraphCommandsRegistrar` to `ShaderGraphCommands`, which calls 'RegisterCommandHandlers'
'PreviewCommandsRegistrar' to 'PreviewCommands', which calls 'RegisterCommandHandlers'

And for the overrides, `GraphViewCommandOverridesRegistrar` to `RegisterGraphViewOverrideCommandHandlers`
`BlackboardCommandOverridesRegistrar' to 'RegisterBlackboardOverrideCommandHandlers'

The static class wrappers are really just a convenience thing cause in the case of the custom shader graph commands (and the preview commands) there are enough of them to merit their own file, and also logically they don't seem to make sense anywhere else

The overrides are fewer so I just tucked them where it made sense to me logically (Graph View overrides in the `ShaderGraphView`, Blackboard overrides where we instantiate the blackboard view in the `EditorWindow`) and just put functions around them as opposed to a class wrapper cause there aren't nearly as many

**Pull request merged**